In [1]:
import gensim
import numpy as np

In [2]:
import gensim.downloader as api
wv = api.load("word2vec-google-news-300")

vec_king = wv["king"]

In [3]:
vec_king

array([ 1.25976562e-01,  2.97851562e-02,  8.60595703e-03,  1.39648438e-01,
       -2.56347656e-02, -3.61328125e-02,  1.11816406e-01, -1.98242188e-01,
        5.12695312e-02,  3.63281250e-01, -2.42187500e-01, -3.02734375e-01,
       -1.77734375e-01, -2.49023438e-02, -1.67968750e-01, -1.69921875e-01,
        3.46679688e-02,  5.21850586e-03,  4.63867188e-02,  1.28906250e-01,
        1.36718750e-01,  1.12792969e-01,  5.95703125e-02,  1.36718750e-01,
        1.01074219e-01, -1.76757812e-01, -2.51953125e-01,  5.98144531e-02,
        3.41796875e-01, -3.11279297e-02,  1.04492188e-01,  6.17675781e-02,
        1.24511719e-01,  4.00390625e-01, -3.22265625e-01,  8.39843750e-02,
        3.90625000e-02,  5.85937500e-03,  7.03125000e-02,  1.72851562e-01,
        1.38671875e-01, -2.31445312e-01,  2.83203125e-01,  1.42578125e-01,
        3.41796875e-01, -2.39257812e-02, -1.09863281e-01,  3.32031250e-02,
       -5.46875000e-02,  1.53198242e-02, -1.62109375e-01,  1.58203125e-01,
       -2.59765625e-01,  

In [4]:
vec_king.shape

(300,)

In [5]:
import re, nltk, pandas as pd
from nltk import WordNetLemmatizer
lm = WordNetLemmatizer()
from nltk.corpus import stopwords
from nltk import sent_tokenize
from gensim.utils import simple_preprocess

In [6]:
msg = pd.read_csv(r"C:\Users\rauna\codes\datasets\SMSSpamCollection", sep ="\t", names = ["label","message"])

In [29]:
msg

label                                            message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [7]:
corpus = []
for i in range(0,len(msg)):
    review = re.sub("[^a-zA-Z]"," ", msg["message"][i])
    review = review.lower()
    review = review.split()
    review = [lm.lemmatize(word) for word in review ]
    review = " ".join(review)
    corpus.append(review)

In [8]:
corpus

['go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat',
 'ok lar joking wif u oni',
 'free entry in a wkly comp to win fa cup final tkts st may text fa to to receive entry question std txt rate t c s apply over s',
 'u dun say so early hor u c already then say',
 'nah i don t think he go to usf he life around here though',
 'freemsg hey there darling it s been week s now and no word back i d like some fun you up for it still tb ok xxx std chgs to send to rcv',
 'even my brother is not like to speak with me they treat me like aid patent',
 'a per your request melle melle oru minnaminunginte nurungu vettam ha been set a your callertune for all caller press to copy your friend callertune',
 'winner a a valued network customer you have been selected to receivea prize reward to claim call claim code kl valid hour only',
 'had your mobile month or more u r entitled to update to the latest colour mobile with camera for free call the mobile up

In [9]:
words = []
for sent in corpus:
    sent_token = sent_tokenize(sent)
    for sent in sent_token:
        words.append(simple_preprocess(sent))

words

[['go',
  'until',
  'jurong',
  'point',
  'crazy',
  'available',
  'only',
  'in',
  'bugis',
  'great',
  'world',
  'la',
  'buffet',
  'cine',
  'there',
  'got',
  'amore',
  'wat'],
 ['ok', 'lar', 'joking', 'wif', 'oni'],
 ['free',
  'entry',
  'in',
  'wkly',
  'comp',
  'to',
  'win',
  'fa',
  'cup',
  'final',
  'tkts',
  'st',
  'may',
  'text',
  'fa',
  'to',
  'to',
  'receive',
  'entry',
  'question',
  'std',
  'txt',
  'rate',
  'apply',
  'over'],
 ['dun', 'say', 'so', 'early', 'hor', 'already', 'then', 'say'],
 ['nah',
  'don',
  'think',
  'he',
  'go',
  'to',
  'usf',
  'he',
  'life',
  'around',
  'here',
  'though'],
 ['freemsg',
  'hey',
  'there',
  'darling',
  'it',
  'been',
  'week',
  'now',
  'and',
  'no',
  'word',
  'back',
  'like',
  'some',
  'fun',
  'you',
  'up',
  'for',
  'it',
  'still',
  'tb',
  'ok',
  'xxx',
  'std',
  'chgs',
  'to',
  'send',
  'to',
  'rcv'],
 ['even',
  'my',
  'brother',
  'is',
  'not',
  'like',
  'to',
  'spea

In [10]:
len(words)

5569

In [11]:
model = gensim.models.Word2Vec(words)

In [12]:
model.wv.index_to_key

['to',
 'you',
 'the',
 'it',
 'and',
 'in',
 'is',
 'me',
 'my',
 'for',
 'your',
 'call',
 'of',
 'that',
 'have',
 'on',
 'now',
 'are',
 'can',
 'so',
 'but',
 'not',
 'or',
 'we',
 'do',
 'get',
 'at',
 'ur',
 'will',
 'if',
 'be',
 'with',
 'no',
 'just',
 'this',
 'gt',
 'lt',
 'go',
 'how',
 'up',
 'when',
 'ok',
 'day',
 'what',
 'free',
 'from',
 'all',
 'out',
 'know',
 'll',
 'come',
 'like',
 'good',
 'time',
 'am',
 'then',
 'got',
 'wa',
 'there',
 'he',
 'love',
 'text',
 'only',
 'want',
 'send',
 'one',
 'need',
 'txt',
 'today',
 'by',
 'going',
 'don',
 'stop',
 'home',
 'she',
 'about',
 'lor',
 'sorry',
 'see',
 'still',
 'mobile',
 'take',
 'back',
 'da',
 'reply',
 'dont',
 'our',
 'think',
 'tell',
 'week',
 'hi',
 'phone',
 'they',
 'new',
 'please',
 'later',
 'pls',
 'any',
 'her',
 'ha',
 'co',
 'did',
 'been',
 'msg',
 'min',
 'some',
 'an',
 'night',
 'make',
 'dear',
 'who',
 'here',
 'message',
 'say',
 'well',
 'where',
 're',
 'thing',
 'much',
 'oh',

In [13]:
model.corpus_count

5569

In [14]:
model.epochs

5

In [15]:
model.wv.similar_by_word("good")

[('my', 0.9989801645278931),
 ('all', 0.9987093806266785),
 ('well', 0.9987028241157532),
 ('day', 0.9986220598220825),
 ('night', 0.9985997080802917),
 ('morning', 0.9985854029655457),
 ('of', 0.998566746711731),
 ('wa', 0.998526930809021),
 ('thing', 0.9984989762306213),
 ('happy', 0.998493492603302)]

In [16]:
model.wv["good"].shape

(100,)

In [17]:
words[0]

['go',
 'until',
 'jurong',
 'point',
 'crazy',
 'available',
 'only',
 'in',
 'bugis',
 'great',
 'world',
 'la',
 'buffet',
 'cine',
 'there',
 'got',
 'amore',
 'wat']

In [18]:
import numpy as np
def avg_word2vec(doc):
    
    return np.mean([model.wv[word] for word in doc if word in model.wv.index_to_key],axis=0)

#We are doing the avgword2vec on every sentence to give single vector for a sentence

In [19]:
#!pip install tqdm

In [20]:
from tqdm import tqdm

In [123]:
x = []
for i in tqdm(range(len(words))):
    x.append(avg_word2vec(words[i]))   
    


  0%|          | 0/5569 [00:00<?, ?it/s]c:\Users\rauna\codes\venv\Lib\site-packages\numpy\core\fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
c:\Users\rauna\codes\venv\Lib\site-packages\numpy\core\_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)
100%|██████████| 5569/5569 [00:00<00:00, 5670.84it/s]


In [124]:
x

[array([-0.16488509,  0.2458592 ,  0.13147846,  0.08058992,  0.09035875,
        -0.4903913 ,  0.19047569,  0.46645907, -0.27588424, -0.13607281,
        -0.14976549, -0.36121464, -0.05081129,  0.08712117,  0.19835916,
        -0.16641419,  0.11920949, -0.3010318 , -0.08621991, -0.5017623 ,
         0.19080211,  0.08891479,  0.07271603, -0.23232694, -0.04036279,
        -0.02940705, -0.20546857, -0.19124886, -0.27750152,  0.02217591,
         0.29876086,  0.01542409,  0.10771557, -0.18223934, -0.11470225,
         0.38340178,  0.08455712, -0.11062834, -0.11942474, -0.45238316,
         0.09516092, -0.22303429, -0.17369352, -0.00240461,  0.14645866,
         0.02392968, -0.09805039, -0.02453258,  0.21395521,  0.13937782,
         0.15813352, -0.19390245, -0.04330641,  0.05484362, -0.05885464,
         0.06403838,  0.15495609,  0.02323537, -0.34561357,  0.17950241,
         0.01258718,  0.14632978, -0.00495854, -0.11524419, -0.28701395,
         0.26601228,  0.090928  ,  0.21872403, -0.3

In [125]:
len(x)

5569

In [126]:
x_new = np.array(x, dtype= object)

In [127]:
x_new.shape

(5569,)

In [128]:
msg.shape

(5572, 2)

where does 3 records go??

In [129]:
[[i,j,k] for i,j,k in zip(list(map(len,corpus)), corpus, msg["message"]) if i<1]

[[0, '', '645'], [0, '', ':) '], [0, '', ':-) :-)']]

review = re.sub("[^a-zA-Z]"," ", msg["message"][i])
- This removes everything except the alphabets so that why above things get removed.

[[0, '', '645'], [0, '', ':) '], [0, '', ':-) :-)']]
- here, the last term is what get removed

In [130]:
x_new[0]

array([-0.16488509,  0.2458592 ,  0.13147846,  0.08058992,  0.09035875,
       -0.4903913 ,  0.19047569,  0.46645907, -0.27588424, -0.13607281,
       -0.14976549, -0.36121464, -0.05081129,  0.08712117,  0.19835916,
       -0.16641419,  0.11920949, -0.3010318 , -0.08621991, -0.5017623 ,
        0.19080211,  0.08891479,  0.07271603, -0.23232694, -0.04036279,
       -0.02940705, -0.20546857, -0.19124886, -0.27750152,  0.02217591,
        0.29876086,  0.01542409,  0.10771557, -0.18223934, -0.11470225,
        0.38340178,  0.08455712, -0.11062834, -0.11942474, -0.45238316,
        0.09516092, -0.22303429, -0.17369352, -0.00240461,  0.14645866,
        0.02392968, -0.09805039, -0.02453258,  0.21395521,  0.13937782,
        0.15813352, -0.19390245, -0.04330641,  0.05484362, -0.05885464,
        0.06403838,  0.15495609,  0.02323537, -0.34561357,  0.17950241,
        0.01258718,  0.14632978, -0.00495854, -0.11524419, -0.28701395,
        0.26601228,  0.090928  ,  0.21872403, -0.32844275,  0.39

In [131]:
y =msg[list(map(lambda x: len(x)> 0, corpus))]
y = pd.get_dummies(y["label"],dtype=int)
y = y.iloc[:,0].values

In [132]:
y

array([1, 1, 0, ..., 1, 1, 1])

In [133]:
y.shape

(5569,)

In [134]:
x[0].shape

(100,)

In [135]:
x[0].reshape(1,-1).shape  # one row, 100 columns

(1, 100)

In [136]:
# More efficient approach using pd.concat
df_list = []
for i in range(0, len(x)):
    df_list.append(pd.DataFrame(x[i].reshape(1, -1)))

df = pd.concat(df_list, ignore_index=True)

C:\Users\rauna\AppData\Local\Temp\ipykernel_25552\4271134234.py:6: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  df = pd.concat(df_list, ignore_index=True)


In [137]:
df

0         1         2         3         4         5         6   \
0    -0.164885  0.245859  0.131478  0.080590  0.090359 -0.490391  0.190476   
1    -0.159066  0.218976  0.113147  0.070510  0.087851 -0.436204  0.155200   
2    -0.175033  0.264611  0.153115  0.090081  0.073472 -0.543685  0.199480   
3    -0.232708  0.333722  0.169112  0.114796  0.124764 -0.663088  0.249064   
4    -0.203904  0.279423  0.152508  0.096027  0.114713 -0.564242  0.211016   
...        ...       ...       ...       ...       ...       ...       ...   
5564 -0.198719  0.307896  0.182242  0.105188  0.098344 -0.611704  0.237408   
5565 -0.210338  0.297628  0.160065  0.107480  0.110045 -0.600531  0.229944   
5566 -0.230629  0.337378  0.171609  0.102115  0.115502 -0.656222  0.258318   
5567 -0.206805  0.302303  0.161351  0.099762  0.106309 -0.603632  0.228792   
5568 -0.234474  0.322121  0.171906  0.113639  0.122753 -0.650700  0.240698   

            7         8         9   ...        90        91        92  \
0     0.466459 -0.275884 -0.136073  ...  0.358061  0.145976  0.032308   
1     0.419850 -0.246543 -0.114765  ...  0.324486  0.122141  0.020874   
2     0.463582 -0.297229 -0.169178  ...  0.354809  0.151580  0.023656   
3     0.641742 -0.379617 -0.176399  ...  0.492591  0.193260  0.039661   
4     0.552002 -0.325520 -0.155463  ...  0.426654  0.162572  0.039927   
...        ...       ...       ...  ...       ...       ...       ...   
5564  0.551811 -0.343647 -0.183343  ...  0.428495  0.180706  0.038377   
5565  0.567433 -0.351760 -0.166829  ...  0.440557  0.181163  0.025124   
5566  0.640467 -0.373271 -0.187772  ...  0.492586  0.204493  0.052666   
5567  0.572898 -0.339651 -0.174548  ...  0.436601  0.179634  0.046164   
5568  0.623355 -0.368363 -0.188523  ...  0.472458  0.189437  0.045519   

            93        94        95        96        97        98        99  
0     0.050826  0.440888  0.184501  0.127934 -0.209416  0.159572 -0.006460  
1     0.040247  0.379843  0.157929  0.112122 -0.196425  0.148178 -0.013320  
2     0.029101  0.462050  0.177761  0.086447 -0.244169  0.184585  0.017218  
3     0.077300  0.587513  0.254563  0.188517 -0.289371  0.218320 -0.020130  
4     0.069070  0.504796  0.218445  0.156732 -0.255179  0.179610 -0.018752  
...        ...       ...       ...       ...       ...       ...       ...  
5564  0.042264  0.543355  0.210925  0.117024 -0.271212  0.202126  0.014828  
5565  0.056725  0.531782  0.218732  0.139177 -0.273081  0.199985 -0.003685  
5566  0.083496  0.596298  0.259063  0.191157 -0.275355  0.207639 -0.016748  
5567  0.066366  0.539865  0.229144  0.152193 -0.260422  0.190698 -0.012700  
5568  0.069736  0.580674  0.250536  0.164477 -0.290219  0.204995 -0.012791  

[5569 rows x 100 columns]

In [138]:
df["output"] = y
df

0         1         2         3         4         5         6  \
0    -0.164885  0.245859  0.131478  0.080590  0.090359 -0.490391  0.190476   
1    -0.159066  0.218976  0.113147  0.070510  0.087851 -0.436204  0.155200   
2    -0.175033  0.264611  0.153115  0.090081  0.073472 -0.543685  0.199480   
3    -0.232708  0.333722  0.169112  0.114796  0.124764 -0.663088  0.249064   
4    -0.203904  0.279423  0.152508  0.096027  0.114713 -0.564242  0.211016   
...        ...       ...       ...       ...       ...       ...       ...   
5564 -0.198719  0.307896  0.182242  0.105188  0.098344 -0.611704  0.237408   
5565 -0.210338  0.297628  0.160065  0.107480  0.110045 -0.600531  0.229944   
5566 -0.230629  0.337378  0.171609  0.102115  0.115502 -0.656222  0.258318   
5567 -0.206805  0.302303  0.161351  0.099762  0.106309 -0.603632  0.228792   
5568 -0.234474  0.322121  0.171906  0.113639  0.122753 -0.650700  0.240698   

             7         8         9  ...        91        92        93  \
0     0.466459 -0.275884 -0.136073  ...  0.145976  0.032308  0.050826   
1     0.419850 -0.246543 -0.114765  ...  0.122141  0.020874  0.040247   
2     0.463582 -0.297229 -0.169178  ...  0.151580  0.023656  0.029101   
3     0.641742 -0.379617 -0.176399  ...  0.193260  0.039661  0.077300   
4     0.552002 -0.325520 -0.155463  ...  0.162572  0.039927  0.069070   
...        ...       ...       ...  ...       ...       ...       ...   
5564  0.551811 -0.343647 -0.183343  ...  0.180706  0.038377  0.042264   
5565  0.567433 -0.351760 -0.166829  ...  0.181163  0.025124  0.056725   
5566  0.640467 -0.373271 -0.187772  ...  0.204493  0.052666  0.083496   
5567  0.572898 -0.339651 -0.174548  ...  0.179634  0.046164  0.066366   
5568  0.623355 -0.368363 -0.188523  ...  0.189437  0.045519  0.069736   

            94        95        96        97        98        99  output  
0     0.440888  0.184501  0.127934 -0.209416  0.159572 -0.006460       1  
1     0.379843  0.157929  0.112122 -0.196425  0.148178 -0.013320       1  
2     0.462050  0.177761  0.086447 -0.244169  0.184585  0.017218       0  
3     0.587513  0.254563  0.188517 -0.289371  0.218320 -0.020130       1  
4     0.504796  0.218445  0.156732 -0.255179  0.179610 -0.018752       1  
...        ...       ...       ...       ...       ...       ...     ...  
5564  0.543355  0.210925  0.117024 -0.271212  0.202126  0.014828       0  
5565  0.531782  0.218732  0.139177 -0.273081  0.199985 -0.003685       1  
5566  0.596298  0.259063  0.191157 -0.275355  0.207639 -0.016748       1  
5567  0.539865  0.229144  0.152193 -0.260422  0.190698 -0.012700       1  
5568  0.580674  0.250536  0.164477 -0.290219  0.204995 -0.012791       1  

[5569 rows x 101 columns]

In [139]:
df.dropna(inplace=True)


In [154]:
x= df.drop("output", axis=1)

In [155]:
x.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
95    0
96    0
97    0
98    0
99    0
Length: 100, dtype: int64

In [156]:
y = df["output"]

In [157]:
y

0       1
1       1
2       0
3       1
4       1
       ..
5564    0
5565    1
5566    1
5567    1
5568    1
Name: output, Length: 5557, dtype: int32

In [158]:
from sklearn.model_selection import train_test_split
x_train,x_test, y_train, y_test = train_test_split(x,y, test_size=0.3)

In [159]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()

In [160]:
x

0         1         2         3         4         5         6  \
0    -0.164885  0.245859  0.131478  0.080590  0.090359 -0.490391  0.190476   
1    -0.159066  0.218976  0.113147  0.070510  0.087851 -0.436204  0.155200   
2    -0.175033  0.264611  0.153115  0.090081  0.073472 -0.543685  0.199480   
3    -0.232708  0.333722  0.169112  0.114796  0.124764 -0.663088  0.249064   
4    -0.203904  0.279423  0.152508  0.096027  0.114713 -0.564242  0.211016   
...        ...       ...       ...       ...       ...       ...       ...   
5564 -0.198719  0.307896  0.182242  0.105188  0.098344 -0.611704  0.237408   
5565 -0.210338  0.297628  0.160065  0.107480  0.110045 -0.600531  0.229944   
5566 -0.230629  0.337378  0.171609  0.102115  0.115502 -0.656222  0.258318   
5567 -0.206805  0.302303  0.161351  0.099762  0.106309 -0.603632  0.228792   
5568 -0.234474  0.322121  0.171906  0.113639  0.122753 -0.650700  0.240698   

             7         8         9  ...        90        91        92  \
0     0.466459 -0.275884 -0.136073  ...  0.358061  0.145976  0.032308   
1     0.419850 -0.246543 -0.114765  ...  0.324486  0.122141  0.020874   
2     0.463582 -0.297229 -0.169178  ...  0.354809  0.151580  0.023656   
3     0.641742 -0.379617 -0.176399  ...  0.492591  0.193260  0.039661   
4     0.552002 -0.325520 -0.155463  ...  0.426654  0.162572  0.039927   
...        ...       ...       ...  ...       ...       ...       ...   
5564  0.551811 -0.343647 -0.183343  ...  0.428495  0.180706  0.038377   
5565  0.567433 -0.351760 -0.166829  ...  0.440557  0.181163  0.025124   
5566  0.640467 -0.373271 -0.187772  ...  0.492586  0.204493  0.052666   
5567  0.572898 -0.339651 -0.174548  ...  0.436601  0.179634  0.046164   
5568  0.623355 -0.368363 -0.188523  ...  0.472458  0.189437  0.045519   

            93        94        95        96        97        98        99  
0     0.050826  0.440888  0.184501  0.127934 -0.209416  0.159572 -0.006460  
1     0.040247  0.379843  0.157929  0.112122 -0.196425  0.148178 -0.013320  
2     0.029101  0.462050  0.177761  0.086447 -0.244169  0.184585  0.017218  
3     0.077300  0.587513  0.254563  0.188517 -0.289371  0.218320 -0.020130  
4     0.069070  0.504796  0.218445  0.156732 -0.255179  0.179610 -0.018752  
...        ...       ...       ...       ...       ...       ...       ...  
5564  0.042264  0.543355  0.210925  0.117024 -0.271212  0.202126  0.014828  
5565  0.056725  0.531782  0.218732  0.139177 -0.273081  0.199985 -0.003685  
5566  0.083496  0.596298  0.259063  0.191157 -0.275355  0.207639 -0.016748  
5567  0.066366  0.539865  0.229144  0.152193 -0.260422  0.190698 -0.012700  
5568  0.069736  0.580674  0.250536  0.164477 -0.290219  0.204995 -0.012791  

[5557 rows x 100 columns]

In [161]:
y

0       1
1       1
2       0
3       1
4       1
       ..
5564    0
5565    1
5566    1
5567    1
5568    1
Name: output, Length: 5557, dtype: int32

In [162]:
rf.fit(x_train,y_train)

RandomForestClassifier()

In [163]:
y_pred = rf.predict(x_test)

In [164]:
from sklearn.metrics import accuracy_score,classification_report, confusion_matrix

In [165]:
print(accuracy_score(y_test,y_pred))

0.9772182254196643


In [166]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.95      0.88      0.91       219
           1       0.98      0.99      0.99      1449

    accuracy                           0.98      1668
   macro avg       0.96      0.93      0.95      1668
weighted avg       0.98      0.98      0.98      1668



In [167]:
print(confusion_matrix(y_test,y_pred))

[[ 192   27]
 [  11 1438]]
